apps/sms/test/marionette/lib/messages.js

```diff
@@ -256,6 +256,34 @@
           client.helper.waitForElement(SELECTORS.main);
         },
 
+        launchFromNotification: function(notification) {
+          var chromeClient = client.scope({ context: 'chrome' });
+          chromeClient.executeScript(function(origin, notification) {
+            /* global Components, Services */
+            var managerClass = Components.classes[
+              '@mozilla.org/system-message-internal;1'
+            ];
+
+            var pageURI = Services.io.newURI(
+              origin + '/index.html#notification', null, null
+            );
+
+            var manifestURI = Services.io.newURI(
+              origin + '/manifest.webapp', null, null
+            );
+
+            var systemMessageManager = managerClass.getService(
+              Components.interfaces.nsISystemMessagesInternal
+            );
+
+            systemMessageManager.sendMessage(
+              'notification', notification, pageURI, manifestURI
+            );
+          }, [ORIGIN_URL, notification]);
+
+          this.switchTo();
+        },
+
```

Comment:
> it can change and it definitely will change and our app (and test) should not depend on it.

That's why I said we should reuse the lib that is in the system app -- hopefully this lib will change if the notification subsystem changes.

However you're currently calling internal gecko modules. I don't think it's more future-proof. I agree with your later statement that this should be made easier.

> feels more like gaia-integration test rather than sms integration

Reminds me of my discussion with Johan :) I can tell you discuss with him a lot these days ;)

Currently I don't really see the point of such integration tests; I think end to end integration tests have more value (testing user scenarios), and that more fine-grained tests can be done with unit tests.

But I don't want to block on this; it's still better to have this tested than not.

My main issue here is that we hardcode pageURI; from what I see in Gecko code, the piece of code to retrieve pageURI is not really shared. For example for notifications it's in https://dxr.mozilla.org/mozilla-central/source/b2g/components/AlertsHelper.jsm#185-202, but this code is not used elsewhere: https://dxr.mozilla.org/mozilla-central/source/dom/apps/Webapps.jsm#872-892 https://dxr.mozilla.org/mozilla-central/source/dom/apps/AppsUtils.jsm#458-467 

However it seems like that we can forget pageURI and it should work: https://dxr.mozilla.org/mozilla-central/source/dom/messages/SystemMessageInternal.js#229-238

> later we can add "one with keeping it" test as well.

I'm fine with keeping this second test for later.

>  ideally marionette should provide smth like launchFromSystemMessage since it's declared in app manifest

I don't think "launch" is really part of marionnette, it's part of the marionette js client library: https://github.com/mozilla-b2g/marionette-apps

You're right that we could try to augment it, at least this would be a more centralized way of making it work.